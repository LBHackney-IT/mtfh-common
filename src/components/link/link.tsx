import React, { ComponentPropsWithoutRef, forwardRef } from 'react';
import cn from 'classnames';
import { navigateToUrl } from 'single-spa';

import './styles.scss';

export interface LinkProps extends ComponentPropsWithoutRef<'a'> {
  variant?:
    | 'link'
    | 'danger'
    | 'text-colour'
    | 'muted'
    | 'back-link'
    | 'native';
  isExternal?: boolean;
}

export const Link = forwardRef<HTMLAnchorElement, LinkProps>(function Link(
  {
    variant = 'link',
    isExternal = false,
    href,
    className,
    children,
    rel,
    target,
    ...props
  },
  ref
) {
  const linkClasses = cn(
    variant !== 'native' && {
      'govuk-link lbh-link': variant !== 'back-link',
      'govuk-back-link lbh-back-link': variant === 'back-link',
      [`lbh-link--${variant}`]: variant !== 'link' && variant !== 'back-link',
    },
    className
  );
  return (
    // eslint-disable-next-line react/jsx-no-target-blank
    <a
      ref={ref}
      href={href}
      className={linkClasses}
      rel={isExternal ? 'noopener noreferrer' : rel}
      target={isExternal ? '_blank' : target}
      onClick={!isExternal ? navigateToUrl : undefined}
      {...props}
    >
      {children}
    </a>
  );
});
