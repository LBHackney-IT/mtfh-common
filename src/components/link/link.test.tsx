import React, { useEffect } from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { navigateToUrl } from 'single-spa';

import { Link } from './link';

jest.mock('single-spa');

test('it renders correctly', () => {
  const { container } = render(
    <Link href="http://localhost">Test Button</Link>
  );
  expect(container).toMatchSnapshot();
});

test('it renders a variant', () => {
  const { container } = render(
    <Link href="http://localhost" variant="muted">
      Test Button
    </Link>
  );
  expect(container).toMatchSnapshot();
});

test('it renders a back link', () => {
  const { container } = render(
    <Link href="http://localhost" variant="back-link">
      Test Button
    </Link>
  );
  expect(container).toMatchSnapshot();
});

test('it does not use the navigation when marked as external', () => {
  (navigateToUrl as jest.Mock).mockReset();
  render(
    <Link href="http://localhost" isExternal>
      Test Button
    </Link>
  );
  const link = screen.getByText('Test Button');
  fireEvent.click(link);
  expect(navigateToUrl).toHaveBeenCalledTimes(0);
});

test('it accepts a ref', () => {
  const callback = jest.fn();
  const Comp = () => {
    const ref = React.createRef<HTMLAnchorElement>();

    useEffect(() => {
      if (ref.current) {
        callback(ref.current);
      }
    }, [ref]);
    return (
      <Link ref={ref} href="http://localhost">
        Test Button
      </Link>
    );
  };
  render(<Comp />);
  expect(callback).toBeCalledTimes(1);
});
