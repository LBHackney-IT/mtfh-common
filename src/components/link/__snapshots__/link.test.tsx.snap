// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a back link 1`] = `
<div>
  <a
    class="govuk-back-link lbh-back-link"
    href="http://localhost"
  >
    Test Button
  </a>
</div>
`;

exports[`it renders a variant 1`] = `
<div>
  <a
    class="govuk-link lbh-link lbh-link--muted"
    href="http://localhost"
  >
    Test Button
  </a>
</div>
`;

exports[`it renders correctly 1`] = `
<div>
  <a
    class="govuk-link lbh-link"
    href="http://localhost"
  >
    Test Button
  </a>
</div>
`;
